src/sklearn_raster/ufunc.py

```diff
@@ -0,0 +1,293 @@
+            is fully masked and `skip_nodata=True`, dummy values (`nan_fill` or 0) will
+            be inserted to ensure this number of samples. This is necessary for
+            functions that require non-empty array inputs, like some scikit-learn
+            `predict` methods. No effect if the array contains enough valid pixels or if
+            `skip_nodata=False`.
+        allow_cast : bool, default False
+            If True and the `func` output dtype is incompatible with the chosen
+            `nodata_output` value, the output will be cast to the correct dtype.
+            Otherwise, an error will be raised.
+        check_output_for_nodata : bool, default True
+            If True and `nodata_output` is not np.nan, a warning will be raised if the
+            selected `nodata_output` value is returned by `func`, as this may indicate
+            a valid pixel being masked.
+        **kwargs : dict
+            Additional keyword arguments passed to `func`.
+        """
+        # Fill NaNs in the input array if they're not being skipped
+        flat_array = self.flat_array if skip_nodata else self._fill_flat_nans(nan_fill)
+        # Only skip NoData if there's something to skip
+        if skip_nodata and self._num_masked > 0:
+            flat_result = self._skip_nodata_apply(
+                func,
+                flat_array=flat_array,
+                ensure_min_samples=ensure_min_samples,
+                nodata_output=nodata_output,
+                allow_cast=allow_cast,
+                nan_fill=nan_fill,
+                check_output_for_nodata=check_output_for_nodata,
+                **kwargs,
+            )
+            # NoData is now pre-masked
+            mask_nodata = False
+        else:
+            flat_result = func(flat_array, **kwargs)
+            mask_nodata = self._num_masked > 0
+
+        @map_function_over_tuples
+        def _unflatten_and_mask(result: NDArray) -> NDArray:
+            """Unflattening result to (y, x, band) and mask NoData."""
+            output_shape = [*self.array.shape[:2], -1]
+            if mask_nodata:
+                result = self._mask_nodata(
+                    result,
+                    nodata_output=nodata_output,
+                    allow_cast=allow_cast,
+                    check_output_for_nodata=check_output_for_nodata,
+                )
+
+            return result.reshape(output_shape)
+
+        return _unflatten_and_mask(flat_result)
+
+    def _skip_nodata_apply(
+        self,
+        func: ArrayUfunc,
+        *,
+        flat_array: NDArray,
+        nodata_output: float | int,
+        ensure_min_samples: int,
+        allow_cast: bool,
+        nan_fill: float | int | None,
+        check_output_for_nodata: bool,
+        **kwargs,
+    ) -> NDArray | tuple[NDArray, ...]:
+        """
+        Apply a function to all non-NoData values in a flat array.
+        """
```

Comment:
Absolute nit.  One-liner docstrings not consistent throughout this module in terms of how they are formatted (quotes on separate lines.)

Reply:
Thanks! I'll fix this and `_mask_nodata`.